File: nn-2opt.py
import math, sys, time

def run_tsp(graph):
	# Make a copy of the graph in order to store the unvisited cities
	unvisited_cities = dict(graph)

	# Start at the first city and remove it from the unvisited cities
	current_city = list(unvisited_cities.keys())[0]
	del unvisited_cities[current_city]

	# Add the first city to the tour
	tour = [current_city]

	# Visit every city in order to get the tour
	while len(unvisited_cities):
		min_distance = float('inf')

		# Check all of the unvisited cities which neighbor the current city
		for neighbor_city in unvisited_cities:
			# Get the distance from the current city to the neighboring city
			neighbor_distance = round(math.hypot(graph[current_city][0] - unvisited_cities[neighbor_city][0],
				graph[current_city][1] - unvisited_cities[neighbor_city][1]))

			# Store information about the neighboring city if it is closer than the last checked neighboring city
			if neighbor_distance < min_distance:
				closest_city = neighbor_city
				min_distance = neighbor_distance

		# Update the current city and remove the closest city from the unvisited cities
		current_city = closest_city
		del unvisited_cities[closest_city]

		# Add the nearest neighboring city to the tour
		tour.append(closest_city)

	return tour

def run_2opt(graph, tour):
	# Make a copy of the tour
	best_tour = tour[:]

	# Set the improvement value
	improvement = True

	# Continuously run 2-opt on the tour until no improvement can be made
	while improvement:
		# Reset the improvement value
		improvement = False

		# Run 2-opt on every possible combination of cities
		for i in range(1, len(best_tour)):
			# Reset the best difference and endpoints
			best_delta = 0
			best_endpoints = [0, 0]

			for j in range(i + 1, len(best_tour) - 1):
				# Calculate the original endpoint distances and swapped endpoint distances
				delta = round((math.hypot(graph[best_tour[i - 1]][0] - graph[best_tour[j]][0], graph[best_tour[i - 1]][1] - graph[best_tour[j]][1])
					+ math.hypot(graph[best_tour[i]][0] - graph[best_tour[j + 1]][0], graph[best_tour[i]][1] - graph[best_tour[j + 1]][1])
					- math.hypot(graph[best_tour[i - 1]][0] - graph[best_tour[i]][0], graph[best_tour[i - 1]][1] - graph[best_tour[i]][1])
					- math.hypot(graph[best_tour[j]][0] - graph[best_tour[j + 1]][0], graph[best_tour[j]][1] - graph[best_tour[j + 1]][1])))

				# Store information about the endpoints if swapping them results in a shorter tour length
				if delta < best_delta:
					best_delta = delta
					best_endpoints = [i, j]

					# Indicate that an improvement has been made
					improvement = True

			# Swap the endpoints if necessary
			if best_delta < 0:
				best_tour[best_endpoints[0]:best_endpoints[1] + 1] = best_tour[best_endpoints[0]:best_endpoints[1] + 1][::-1]

				# Start 2-opt again
				break

	return best_tour

def get_tour_length(graph, tour):
	tour_length = 0

	# Calculate the tour length and round it to the nearest integer
	for i in range(len(tour)):
		tour_length += int(round(math.hypot(graph[tour[i - 1]][0] - graph[tour[i]][0], graph[tour[i - 1]][1] - graph[tour[i]][1])))

	return tour_length

def main(instance_file):
	# Record the start time
	start_time = time.time()

	graph = {}

	# Read the file and store the information in the graph
	with open(instance_file) as f:
		for line in f:
			id, x, y = (int(i) for i in line.split())
			graph[id] = x, y

	# Run the TSP algorithm on the graph to get a tour
	tour = run_tsp(graph)

	# Run the 2-opt algorithm on the tour
	best_tour = run_2opt(graph, tour)

	# Calculate the new tour length
	best_tour_length = get_tour_length(graph, best_tour)

	# Write the results to the file
	with open("{}.tour".format(instance_file), "w") as f:
		f.write("{}\n".format(best_tour_length))
		for i in best_tour:
			f.write("{}\n".format(i))

	# Print the end time
	print "Running time: {}".format(time.time() - start_time)

if __name__ == '__main__':
	main(sys.argv[1])

File: nn-2opt.cpp
#include <chrono>
#include <cmath>
#include <fstream>
#include <iostream>
#include <limits>
#include <map>
#include <string>
#include <vector>

std::vector<int> nearestNeighbor(std::map<int, std::pair<int, int>> &cities, std::vector<std::vector<int>> &distanceMatrix)
{
	// Start at the first city
	int currentCity = (*cities.begin()).first;
	std::vector<int> tour;

	// Remove the first city from the unvisited cities and add it to the tour
	cities.erase(currentCity);
	tour.push_back(currentCity);

	// Visit every city in order to get the tour
	while (!cities.empty())
	{
		int closestCity = 0;
		int minimumDistance = std::numeric_limits<int>::max();

		// Check all of the unvisited cities which neighbor the current city
		for (std::map<int, std::pair<int, int>>::iterator i = cities.begin(); i != cities.end(); ++i)
		{
			// Get the distance from the current city to the neighboring city
			int distance = distanceMatrix[i->first][currentCity];

			// Store information about the neighboring city if it is closer than the last checked neighboring city
			if (distance < minimumDistance)
			{
				closestCity = i->first;
				minimumDistance = distance;
			}
		}

		// Remove the closest city from the unvisited cities and add it to the tour
		cities.erase(closestCity);
		tour.push_back(closestCity);

		// Make the current city the closest city
		currentCity = closestCity;
	}

	return tour;
}

void twoOpt(std::vector<int> &tour, std::vector<std::vector<int>> &distanceMatrix)
{
	bool improvement = true;

	// Continuously run 2-opt on the tour until no improvement can be made
	while (improvement)
	{
		improvement = false;

		// Run 2-opt on every possible combination of cities
		for (int i = 0; i < tour.size(); ++i)
		{
			int bestDelta = 0;
			std::pair<int, int> bestEndpoints;

			for (int j = i + 1; j < tour.size(); ++j)
			{
				// Skip adjacent cities
				if (tour.size() - j - i == 1)
				{
					continue;
				}

				// Calculate the difference between the original endpoint distances and swapped endpoint distances
				int delta = distanceMatrix[tour[((i - 1) + tour.size()) % tour.size()]][tour[j]]
					+ distanceMatrix[tour[i]][tour[(j + 1) % tour.size()]]
					- distanceMatrix[tour[((i - 1) + tour.size()) % tour.size()]][tour[i]]
					- distanceMatrix[tour[j]][tour[(j + 1) % tour.size()]];

				// Store information about the endpoints if swapping them results in a shorter tour length
				if (delta < bestDelta)
				{
					bestDelta = delta;
					bestEndpoints = std::make_pair(i, j);
					improvement = true;
				}
			}

			if (bestDelta < 0)
			{
				int m = bestEndpoints.first, n = bestEndpoints.second;

				// Swap the endpoints
				while (n > m)
				{
					int cityId = tour[m];
					tour[m++] = tour[n];
					tour[n--] = cityId;
				}

				// Start 2-opt again
				break;
			}
		}
	}
}

std::vector<std::vector<int>> buildDistanceMatrix(std::map<int, std::pair<int, int>> &cities)
{
	std::vector<std::vector<int>> distanceMatrix(cities.size());

	// Calculate and store the distances between each city
	for (std::map<int, std::pair<int, int>>::iterator i = cities.begin(); i != cities.end(); ++i)
	{
		distanceMatrix[i->first] = std::vector<int>(cities.size());

		for (std::map<int, std::pair<int, int>>::iterator j = cities.begin(); j != cities.end(); ++j)
		{
			distanceMatrix[i->first][j->first] = std::lround(std::hypot(i->second.first - j->second.first, i->second.second - j->second.second));
		}
	}

	return distanceMatrix;
}

int calculateTourLength(std::vector<int> &tour, std::vector<std::vector<int>> &distanceMatrix)
{
	int tourLength = 0;

	// Sum the distances between each city
	for (int i = 0; i < tour.size(); ++i)
	{
		tourLength += distanceMatrix[tour[i]][tour[((i + 1) + tour.size()) % tour.size()]];
	}

	return tourLength;
}

std::map<int, std::pair<int, int>> readFile(std::string name)
{
	std::map<int, std::pair<int, int>> cities;
	std::ifstream file(name);

	if (!file)
	{
		std::cout << "Error opening " << name << " for reading" << std::endl;
		exit(1);
	}
	else
	{
		int cityId = 0, cityX = 0, cityY = 0;

		// Read the numbers from the file and store them
		while (file >> cityId >> cityX >> cityY)
		{
			cities.insert(std::make_pair(cityId, std::make_pair(cityX, cityY)));
		}
	}

	return cities;
}

void writeFile(std::string name, std::vector<int> &tour, std::vector<std::vector<int>> &distanceMatrix)
{
	std::ofstream file(name);

	if (!file)
	{
		std::cout << "Error opening " << name << " for writing" << std::endl;
		exit(1);
	}
	else
	{
		// Write the tour length to the file
		file << calculateTourLength(tour, distanceMatrix) << "\n";

		// Write the cities in the tour to the file
		for (std::vector<int>::iterator i = tour.begin(); i != tour.end(); ++i)
		{
			file << (*i) << "\n";
		}
	}
}

int main(int argc, char **argv)
{
	if (argc != 2)
	{
		std::cout << "Invalid number of arguments" << std::endl;
		exit(1);
	}
	else
	{
		// Record the start time
		std::chrono::high_resolution_clock::time_point startTime = std::chrono::high_resolution_clock::now();

		// Read the file to get the cities
		std::map<int, std::pair<int, int>> cities = readFile(std::string(argv[1]));

		// Calculate the distance matrix for the tour
		std::vector<std::vector<int>> distanceMatrix = buildDistanceMatrix(cities);

		// Run the nearest neighbor algorithm to get a tour
		std::vector<int> tour = nearestNeighbor(cities, distanceMatrix);

		// Run the 2-opt algorithm to optimize the tour
		twoOpt(tour, distanceMatrix);

		// Write the tour information to the file
		writeFile(std::string(argv[1]).append(".tour"), tour, distanceMatrix);

		// Record the end time and calculate the duration in seconds
		std::chrono::high_resolution_clock::time_point endTime = std::chrono::high_resolution_clock::now();
		double duration = std::chrono::duration<double>(endTime - startTime).count();

		// Output the running time
		std::cout << "Running time: " << duration << " seconds" << std::endl;
	}

	return 0;
}
